File: runtime/lib/math.dart
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

class MathNatives {
  static num pow(num value, num exponent) {
    if (exponent is int) {
      return value.pow(exponent);
    }
    // Double.pow will call exponent.toDouble().
    return value.toDouble().pow(exponent);
  }
  static double random() => _random();
  static double sqrt(num value) => _sqrt(value.toDouble());
  static double sin(num value) => _sin(value.toDouble());
  static double cos(num value) => _cos(value.toDouble());
  static double tan(num value) => _tan(value.toDouble());
  static double acos(num value) => _acos(value.toDouble());
  static double asin(num value) => _asin(value.toDouble());
  static double atan(num value) => _atan(value.toDouble());
  static double atan2(num a, num b) => _atan2(a.toDouble(), b.toDouble());
  static double exp(num value) => _exp(value.toDouble());
  static double log(num value) => _log(value.toDouble());

  static double _random() native "MathNatives_random";
  static double _sqrt(double value) native "MathNatives_sqrt";
  static double _sin(double value) native "MathNatives_sin";
  static double _cos(double value) native "MathNatives_cos";
  static double _tan(double value) native "MathNatives_tan";
  static double _acos(double value) native "MathNatives_acos";
  static double _asin(double value) native "MathNatives_asin";
  static double _atan(double value) native "MathNatives_atan";
  static double _atan2(double a, double b) native "MathNatives_atan2";
  static double _exp(double value) native "MathNatives_exp";
  static double _log(double value) native "MathNatives_log";
}

File: runtime/lib/math.cc
// Copyright (c) 2011, the Dart project authors.  Please see the AUTHORS file
// for details. All rights reserved. Use of this source code is governed by a
// BSD-style license that can be found in the LICENSE file.

#include <ctype.h>  // isspace.

#include "vm/bootstrap_natives.h"

#include "vm/exceptions.h"
#include "vm/native_entry.h"
#include "vm/object.h"
#include "vm/symbols.h"

namespace dart {

DEFINE_NATIVE_ENTRY(Random_initialSeed, 0, 0) {
  return Integer::New(thread->random()->NextUInt64());
}

DEFINE_NATIVE_ENTRY(SecureRandom_getBytes, 0, 1) {
  GET_NON_NULL_NATIVE_ARGUMENT(Smi, count, arguments->NativeArgAt(0));
  const intptr_t n = count.Value();
  ASSERT((n > 0) && (n <= 8));
  uint8_t buffer[8];
  Dart_EntropySource entropy_source = Dart::entropy_source_callback();
  if ((entropy_source == nullptr) || !entropy_source(buffer, n)) {
    const String& error = String::Handle(String::New(
        "No source of cryptographically secure random numbers available."));
    const Array& args = Array::Handle(Array::New(1));
    args.SetAt(0, error);
    Exceptions::ThrowByType(Exceptions::kUnsupported, args);
  }
  uint64_t result = 0;
  for (intptr_t i = 0; i < n; i++) {
    result = (result << 8) | buffer[i];
  }
  return Integer::New(result);
}

}  // namespace dart
